File: 2-DataStructures/5-Queues/queues.py
# Queues definition: a queue is a collection of elements that supports two principle operations: enqueue and dequeue
# First item in, first item out
# FIFO (First In First Out)
# Operations: enqueue, dequeue, is_empty, size

class Queue:
    def __init__(self):
        self.queue = []

    # is_empty: check if the queue is empty. Time complexity: O(1)
    def is_empty(self):
        return self.queue == []

    # enqueue: insert element at the end of the list. Time complexity: O(1)
    def enqueue(self, data):
        self.queue.append(data)

    # dequeue: remove the first element of the list. Time complexity: O(n) - Doubly-linked list would be O(1)
    def dequeue(self):
        if self.is_empty():
            return None

        data = self.queue[0]
        del self.queue[0]
        return data

    # size: get the size of the queue. Time complexity: O(1)
    def size(self):
        return len(self.queue)

    # peek: peek the first element of the list. Time complexity: O(1)
    def peek(self):
        return self.queue[0]

    # print the queue
    def print_queue(self):
        print(self.queue)


# testing
queue = Queue()
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)
queue.print_queue()
print('Size: ', queue.size())
print('Dequeued: ', queue.dequeue())
queue.print_queue()
print('Size: ', queue.size())
print('Peek: ', queue.peek())
queue.print_queue()
print('Size: ', queue.size())
print('Is empty: ', queue.is_empty())
queue.dequeue()
queue.dequeue()
queue.dequeue()
print('Is empty: ', queue.is_empty())

File: 2-DataStructures/4-Queues/queues.js
// queue definition: FIFO (First In First Out)
// queue operations: enqueue, dequeue, peek, length
//

// 1. Create a queue [using Linked List]
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.length = 0;
  }

  peek() {
    return this.first;
  }

  enqueue(value) {
    const newNode = new Node(value);

    if (this.length === 0) {
      this.first = newNode;
      this.last = newNode;
    } else {
      this.last.next = newNode;
      this.last = newNode;
    }
    this.length++;
    return this
  }

  dequeue() {
    if (!this.first) {
      return null;
    }

    if (this.first === this.last) {
      this.last = null;
    }

    this.first = this.first.next;
    this.length--;
    return this;
  }
}

// 2. Create a queue [using Array]
class QueueArrays {
  constructur() {
    this.array = [];
  }

  peek() {
    return this.array[0];
  }

  enqueue(value) {
    this.array.push(value);
    return this;
  }

  dequeue() {
    this.array.pop();
    return this;
  }
}

const myQueue = new Queue();
myQueue.enqueue('Joy');
myQueue.enqueue('Matt');
myQueue.enqueue('Pavel');
myQueue.enqueue('Samir');
console.log(myQueue.peek());
console.log(myQueue.dequeue());
console.log(myQueue.peek());

const myQueueArray = new Queue();
myQueue.enqueue('Joy');
myQueue.enqueue('Matt');
myQueue.enqueue('Pavel');
myQueue.enqueue('Samir');
console.log(myQueue.peek());
console.log(myQueue.dequeue());
console.log(myQueue.peek());